delegate-app/Main.hs

```diff
@@ -1,62 +1,336 @@
+        traverse_ handler elements
+
+-- | run a delegate server
+runDelegateServer ::
+  -- | the configuration of the delegate server
+  DelegateServerConfig ->
+  DelegateTracerT IO ()
+runDelegateServer conf = do
+  trace (Started $ port conf)
+
+  tracer <- askTracer
+
+  eventQueue <- liftIO . atomically $ do
+    q <- newTQueue
+    writeTQueue q Start
+    pure q
+  frontendRequestQueue <- liftIO . atomically $ newTQueue
+  toClientsChannel <- liftIO . atomically $ newTChan
+
+  clientCounter <- liftIO $ newMVar 0
+
+  let workerAction threadSort =
+        withStartStopTrace threadSort $
+          case threadSort of
+            WebsocketThread -> do
+              liftIO $
+                runServer (show $ host conf) (fromIntegral $ port conf) $
+                  runWithTracer' tracer
+                    . websocketsServer
+                      (ping conf)
+                      (freshClientIdGenerator clientCounter)
+                      frontendRequestQueue
+                      toClientsChannel
+            DelegateRunnerThread ->
+              execDelegateRunnerT $
+                runDelegateLogicSteps
+                  (tick conf)
+                  eventQueue
+                  frontendRequestQueue
+                  toClientsChannel
+            QueueAuctionStageThread ->
+              mbQueueAuctionPhases eventQueue toClientsChannel
 
-consumer :: Chan DelegateInput -> DelegateRunnerT IO ()
-consumer delegateInputs = do
-  -- This will block until an event appears in the queue
-  delegateIn <- liftIO $ readChan delegateInputs
-  delegateResponse <- delegateStep delegateIn
-  liftIO $ mbQueueAuctionPhases delegateResponse delegateInputs
-  -- FIXME: send response to client
   liftIO $
-    putStrLn $
-      "Delegate responses for input: " <> show delegateResponse
+    mapConcurrently_
+      (runWithTracer' tracer . workerAction)
+      [ WebsocketThread
+      , QueueAuctionStageThread
+      , DelegateRunnerThread
+      ]
+  where
+    freshClientIdGenerator clientCounter = do
+      v <- takeMVar clientCounter
+      putMVar clientCounter (v + 1)
+      return v
 
-mbQueueAuctionPhases :: [DelegateResponse] -> Chan DelegateInput -> IO ()
-mbQueueAuctionPhases [AuctionSet terms] events = void $ async queueCurrentStage
+-- | start a worker and log it
+withStartStopTrace :: ThreadSort -> DelegateTracerT IO () -> DelegateTracerT IO ()
+withStartStopTrace threadSort action = do
+  trace $ ThreadEvent ThreadStarted threadSort
+  action
+  -- FIXME: use finally or something like that
+  trace $ ThreadEvent ThreadCancelled threadSort
+
+mbQueueAuctionPhases ::
+  TQueue DelegateEvent ->
+  TChan (ClientResponseScope, DelegateResponse) ->
+  DelegateTracerT IO ()
+mbQueueAuctionPhases delegateEvents toClientsChannel = do
+  terms <-
+    liftIO . atomically $
+      awaitForAuctionSetTerms =<< dupTChan toClientsChannel
+  traceQueueEvent terms
+  liftIO $ forever $ queueCurrentStageAndWaitForNext terms
```

Comment:
Why is this `forever`? This eventually reduces to `forever $ pure ()`.

Reply:
You right, I missed that.